master/internal/db/postgres.go

```diff
@@ -1119,13 +1119,37 @@ WHERE id = :id`, setClause(toUpdate)), trial)
 	return nil
 }
 
+// RollbackSearcherEvents rolls back the events for an experiment to the last step with a
+// checkpoint.
+func (db *PgDB) RollbackSearcherEvents(experimentID int) error {
+	_, err := db.sql.Exec(`
+DELETE FROM searcher_events se
+USING (
+    SELECT
+        (se.content->'msg'->'workload'->>'trial_id')::int trial_id,
+        max(CASE WHEN content->'msg'->'workload'->>'kind' = 'CHECKPOINT_MODEL'
+        THEN (se.content->'msg'->'workload'->>'step_id')::int ELSE 0 END) latest_checkpoint_id
+    FROM searcher_events se
+    GROUP BY trial_id
+	) lc
+WHERE experiment_id = $1
+    AND (se.content->'msg'->'workload'->>'trial_id')::int = lc.trial_id
+    AND (se.content->'msg'->'workload'->>'step_id')::int > lc.latest_checkpoint_id;
+	`, experimentID)
+	if err != nil {
+		return errors.Wrapf(err, "error rolling back events for experiment %d", experimentID)
+	}
+	return nil
+}
+
 // RollBackTrial deletes from the database all steps, checkpoints, and validations for the trial
 // that correspond to steps past lastStep.
 func (db *PgDB) RollBackTrial(id int, lastStep int) error {
 	// This delete cascades to checkpoints and validations.
 	_, err := db.sql.Exec(`
 DELETE FROM steps
 WHERE trial_id = $1 AND id > $2
+   OR trial_id = $1 AND id = $2 AND state != 'COMPLETED'
```

Comment:
In the event we checkpoint, then validate but don't finish it, when we rollback the sequencer it will say to roll back to the end up that step (`id > $2`) and this query deletes all steps after that, but we still have a started but not completed validation in the same step (`id = $2`). There isn't really any harm in this but in master restart you see a bunch of `failed to save workload to the database/duplicate val` in `processAssigned` because the validation never got cleared and when we get assigned we start it over and save it. This is reproducible on latest master, too.

Reply:
Ok, cool.